src/cpp/shared_core/FilePath.cpp

```diff
@@ -1291,8 +1291,52 @@ void FilePath::setLastWriteTime(std::time_t in_time) const
 
 Error FilePath::testWritePermissions() const
 {
-   std::shared_ptr<std::ostream> testStream;
-   return openForWrite(testStream, false);
+   try
+   {
+      std::ostream* pResult = nullptr;
+#ifdef _WIN32
+      using namespace boost::iostreams;
+      HANDLE hFile = ::CreateFileW(m_impl->Path.wstring().c_str(),
+                                   FILE_APPEND_DATA,
+                                   0, // exclusive access
+                                   nullptr,
+                                   OPEN_EXISTING,
+                                   0,
+                                   nullptr);
+      if (hFile == INVALID_HANDLE_VALUE)
+      {
+         Error error = LAST_SYSTEM_ERROR();
+         error.addProperty("path", getAbsolutePath());
+         return error;
+      }
+      file_descriptor_sink fd;
+      fd.open(hFile, close_handle);
+      pResult = new boost::iostreams::stream<file_descriptor_sink>(fd);
+#else
+      using std::ios_base;
+      ios_base::openmode flags = ios_base::in | ios_base::out | ios_base::binary;
+      pResult = new std::ofstream(getAbsolutePath().c_str(), flags);
+#endif
+
+      if (!(*pResult))
+      {
+         delete pResult;
+
+         Error error = systemError(boost::system::errc::no_such_file_or_directory, ERROR_LOCATION);
+         error.addProperty("path", getAbsolutePath());
+         return error;
+      }
+   }
+   catch(const std::exception& e)
+   {
+      Error error = systemError(boost::system::errc::io_error,
+                                ERROR_LOCATION);
+      error.addProperty("what", e.what());
+      error.addProperty("path", getAbsolutePath());
+      return error;
+   }
+
```

Comment:
Right now the ostream will be leaked if the function completes successfully. In openForWrite, the ownership is transferred to a smart pointer before the function exits. I think we should delete it before returning Success.

This also opens another question (which applies to the existing implementation of openForWrite) - is it possible for pResult to be initialized when an exception is thrown? In that case, the ostream would be leaked in that case as well. We could avoid this in both places by declaring the pResult outside of the try and deleting it in the catch if it's not null.

Reply:
Alternately in this implementation we can use a local ostream rather than a pointer, which would solve both problems.

Reply:
Thanks for pointing this out. I cleaned up the deletes and updated openForRead as well. Unfortunately because of this line https://github.com/rstudio/rstudio/blob/bugfix/replace-respect-permissions/src/cpp/shared_core/FilePath.cpp#L1317 I think it's cleaner to use a pointer here than a local variable. I did some testing on windows using a local variable and got some strange results.